### PR TITLE
Update progress and error reporting in clang-tidy

### PR DESCRIPTION
**Pull request by 1ntEgr8**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61674
* #61711
* #61689
* **#61672**

This PR adds a progress bar to clang-tidy, and updates how it threads error codes (when run in parallel). The progress bar is disabled on GHA because backspace escape codes are not supported. 

It also adds a `--quiet` flag to the script.

Screenshot of progress bar:
Screen Shot 2021-07-14 at 3 17 11 PM

Differential Revision: [D29763848](https://our.internmc.facebook.com/intern/diff/D29763848)

### CODE REVIEW

**Review by driazati — COMMENTED** [review]
Some random things I found trying this locally:

* `make clang-tidy` doesn't output anything, there's a missing `redirect=False` arg on `shell_cmd()` in `ClangTidy.full`
* the message `Could not find clang-tidy binary` is confusing since I have it in my path, it should probably say it's specifically looking in .clang-tidy-bin
* running the script should say the warnings as it runs instead of just saying `Warning detected` (also I see tons of warnings on master, is that just because we didn't have the clang-tidy version of Werror on until this stack?)
* we should hide output from passed files (and also probably `.strip()` it since I see a bunch of extra newlines when I run `make clang-tidy`), sections like this in the output don't seem that useful

    <details>

    ```
    8782 warnings generated.
    Suppressed 8874 warnings (8782 in non-user code, 92 NOLINT).
    Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.



    5868 warnings generated.
    Suppressed 5869 warnings (5868 in non-user code, 1 NOLINT).
    Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.



    9818 warnings generated.
    Suppressed 9889 warnings (9818 in non-user code, 71 NOLINT).
    Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.



    9818 warnings generated.
    Suppressed 9889 warnings (9818 in non-user code, 71 NOLINT).
    Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.



    8950 warnings generated.
    Suppressed 9036 warnings (8950 in non-user code, 86 NOLINT).
    Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.



    311 warnings generated.
    Suppressed 315 warnings (311 in non-user code, 4 NOLINT).
    Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.

    ```

    </details>

File: tools/linter/clang_tidy/run.py [hunk 1]
```diff
@@ -337,10 +438,9 @@ def run(options: Any) -> int:
         print("No files detected.")
         sys.exit()
 
-    clang_tidy_output = run_clang_tidy(options, line_filters, files)
-    warnings = extract_warnings(
-        clang_tidy_output, base_dir=options.compile_commands_dir
-    )
+    returncode, output = run_clang_tidy(options, line_filters, files)
+    clang_tidy_output = output
```
Comment (janeyx99) [inline tools/linter/clang_tidy/run.py:442]:
```suggestion
    returncode, clang_tidy_output = run_clang_tidy(options, line_filters, files)
```

File: tools/linter/clang_tidy/run.py [hunk 2]
```diff
@@ -116,13 +196,13 @@ def filter_files(files: Iterable[str], file_patterns: Patterns) -> Iterable[str]
...
-            print("{} omitted due to file filters".format(file))
+            print(f"{file} omitted due to file filters")
 
 
 def get_all_files(paths: List[str]) -> List[str]:
     """Returns all files that are tracked by git in the given paths."""
-    output = run_shell_command(["git", "ls-files"] + paths)
-    return output.split("\n")
+    returncode, output = run_shell_command(["git", "ls-files"] + paths)
+    return output.strip().split("\n")
```
Comment (janeyx99) [inline tools/linter/clang_tidy/run.py:205]:
```suggestion
    return output.strip().splitlines()
```
nit

File: tools/linter/clang_tidy/run.py [hunk 3]
```diff
@@ -352,11 +446,15 @@ def run(options: Any) -> int:
...
-    pwd = os.getcwd() + "/"
     if options.dry_run:
         print(clang_tidy_output)
-    for line in clang_tidy_output.splitlines():
-        if line.startswith(pwd):
-            print(line[len(pwd) :])
-
-    return len(warnings.keys())
+    elif returncode != 0:
+        print("[clang-tidy] Warnings detected!")
```
Comment (driazati) [inline tools/linter/clang_tidy/run.py:452]: What's the value of re-printing the warnings? At least on a small change it seems to mostly parrot what it's already said:


```
$ make clang-tidy CHANGED_ONLY=--changed-only                                                             (pytorch) [ 10:19:19 ]
x clang-tidy: Run clang-tidy
Warning detected in torch/csrc/utils/crash_handler.cpp
/data/users/davidriazati/pytorch/torch/csrc/utils/crash_handler.cpp:19:20: error: variable 'minidump_directory' is non-const and globally accessible, consider making it const [cppcoreguidelines-avoid-non-const-global-variables,-warnings-as-errors]
static std::string minidump_directory; 
                   ^

Traceback (most recent call last):
  File "/home/davidriazati/local/miniconda3/envs/pytorch/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/davidriazati/local/miniconda3/envs/pytorch/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/data/users/davidriazati/pytorch/tools/linter/clang_tidy/__main__.py", line 194, in <module>
    main()
  File "/data/users/davidriazati/pytorch/tools/linter/clang_tidy/__main__.py", line 190, in main
    raise RuntimeError("Your code is not clang-tidy clean!")
RuntimeError: Your code is not clang-tidy clean!
[clang-tidy] Warnings detected!
Summary:
torch/csrc/utils/crash_handler.cpp:19:20: error: variable 'minidump_directory' is non-const and globally accessible, consider making it const [cppcoreguidelines-avoid-non-const-global-variables,-warnings-as-errors]

```

File: tools/linter/clang_tidy/__main__.py [hunk 4]
```diff
@@ -185,15 +178,16 @@ def main() -> None:
...
+            + "We provide a custom build of clang-tidy that has additional checks.\n"
+            + "You can install it by running:\n"
+            + "$ python3 tools/linter/install/clang_tidy.py"
         )
         raise RuntimeError(msg)
 
     return_code = run(options)
     if return_code != 0:
-        raise RuntimeError("Warnings found in clang-tidy output!")
+        raise RuntimeError("Your code is not clang-tidy clean!")
```
Comment (driazati) [inline tools/linter/clang_tidy/__main__.py:190]: The Python traceback isn't super useful for a failure in this case imo, I think it'd be better to just exit with an error like mypy/flake8

```suggestion
        sys.exit(returncode)
```

**Review by driazati — APPROVED** [review]
LGTM! Nice UX improvements, a couple output nit comments but otherwise looks good

File: tools/linter/clang_tidy/run.py (hunk 3 above)
Comment (1ntEgr8) [reply tools/linter/clang_tidy/run.py:452]: That's a good point. Perhaps I can aggregate the warnings and output a better summary. Something like
```
[warning-name]: <num-occurrences>
```

Chromium does this

**Comment by 1ntEgr8** [comment]
@1ntEgr8 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29763848).

**Comment by 1ntEgr8** [comment]
@1ntEgr8 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29763848).

**Comment by 1ntEgr8** [comment]
@1ntEgr8 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29763848).

**Pull request closed without merging**